swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,927 @@
+            )) => self
+                .failed_open
+                .push_back(OpenError::NegotiationFailed(info, error)),
+            ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)) => {
+                self.failed_open.push_back(OpenError::Unsupported {
+                    open_info: info,
+                    protocol: self.protocol,
+                })
+            }
+            ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(apply)) => {
+                void::unreachable(apply)
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
+        if let Some(error) = self.failed_open.pop_front() {
+            return Poll::Ready(ConnectionHandlerEvent::Custom(OutEvent::FailedToOpen(
+                error,
+            )));
+        }
+
+        match self.outbound_streams.poll_next_unpin(cx) {
+            Poll::Ready(Some(outbound_done)) => {
+                return Poll::Ready(ConnectionHandlerEvent::Custom(OutEvent::OutboundEmitted(
+                    outbound_done,
+                )));
+            }
+            Poll::Ready(None) => {
+                // Normally, we'd register a waker here but `Connection` polls us anyway again
+                // after calling `inject` on us which is where we'd use the waker.
+            }
+            Poll::Pending => {}
+        };
+
+        match self.inbound_streams.poll_next_unpin(cx) {
+            Poll::Ready(Some(inbound_done)) => {
+                return Poll::Ready(ConnectionHandlerEvent::Custom(OutEvent::InboundEmitted(
+                    inbound_done,
+                )));
+            }
+            Poll::Ready(None) => {
+                // Normally, we'd register a waker here but `Connection` polls us anyway again
+                // after calling `inject` on us which is where we'd use the waker.
+            }
+            Poll::Pending => {}
+        };
+
+        if let Some(outbound_open_info) = self.pending_outbound_streams.pop_front() {
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(
+                    ReadyUpgrade::new(self.protocol),
+                    outbound_open_info,
+                ),
+            });
+        }
+
+        if self.inbound_streams.is_empty()
+            && self.outbound_streams.is_empty()
+            && self.pending_outbound_streams.is_empty()
+        {
+            if self.keep_alive.is_yes() {
+                // TODO: Make timeout configurable
+                self.keep_alive = KeepAlive::Until(Instant::now() + Duration::from_secs(10))
```

Comment:
It feels weird that this grace period is implemented on the level of ConnectionHandler — shouldn’t it be a NetworkBehaviour setting that applies whenever KeepAlive toggles from Yes to No?

Reply:
I am not happy with this either but I couldn't think of a better way to handle this. `keep_alive` is per connection. My initial implementation would have been a `with_keep_alive_timeout` configuration parameter on the builder that defaults to 30s or something.